java/driver/jni/src/main/java/org/apache/arrow/adbc/driver/jni/impl/JniLoader.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.adbc.driver.jni.impl;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.Locale;
+import java.util.Map;
+import org.apache.arrow.adbc.core.AdbcException;
+import org.apache.arrow.util.Preconditions;
+
+public enum JniLoader {
```

Comment:
(design) it's not a common pattern to use an enum class directly for creating a singleton, and not sure if there are benefits outside of it vs using a regular class.

I'm not even sure `INSTANCE` is needed since things are delegated to `NativeAdbc` anyway. It looks to me that `NativeAdbc` could simply have a static initializer invoking some `JniLoader#checkInitialized()` for example which would cause the library to be loaded or `NativeAdbc` class loading to fail otherwise

Reply:
This pattern is actually recommended by Effective Java. But I can remove the singleton, right.

Reply:
Maybe I should check my on copy then :)

Reply:
Ah, the problem is then the "nice" wrapper methods, if merged into `NativeAdbc`, can't use non-standard-library classes anymore, because of the circular dependency issue...

Reply:
For now I'm going to keep this as-is (note that Arrow itself uses the singleton pattern to gate access to JNI methods) and we can revisit if needed